### PR TITLE
[MASTER] Add redis system

### PR DESCRIPTION
**Pull request by HugoPouliquen**

### CODE REVIEW

File: rotp-redis.gemspec [hunk 1]
```diff
@@ -0,0 +1,32 @@
...
+  spec.version       = Rotp::Redis::VERSION
+  spec.authors       = %w(hpouliquen mdouchement)
+  spec.email         = %w(hugo.pouliquen@predicsis.com marc.douchement@predicsis.com)
+  spec.summary       = 'RotpRedis'
+  spec.description   = 'Extension for ROTP'
+  spec.homepage      = 'https://github.com/PredicSis/rotp-redis'
+  spec.license       = 'MIT'
+
+  spec.files         = `git ls-files -z`.split("\x0")
+  spec.executables   = ['rotpredis']
```
Comment (mdouchement) [inline rotp-redis.gemspec]: Remove this line

File: README.md [hunk 2]
```diff
@@ -1 +1,70 @@
...
+
+## Configuration
+- `redis-server`
+- Into your app:
+  - Add gems in Gemfile:
+    ```ruby
+    source 'https://rubygems.org'
+
+    gem 'rotp'
+    gem 'redis'
```
Comment (mdouchement) [inline README.md]: This gem is already required and installed by `rotp-redis`

File: README.md [hunk 3]
```diff
@@ -1 +1,70 @@
...
+
+```ruby
+gem 'rotp-redis'
+```
+
+## Configuration
+- `redis-server`
+- Into your app:
+  - Add gems in Gemfile:
+    ```ruby
```
Comment (mdouchement) [inline README.md]: Rewrite __Add gems in Gemfile__ as:
```ruby
source 'https://rubygems.org'

gem 'rotp'
gem 'rotp-redis'
```

File: spec/spec_helper.rb [hunk 4]
```diff
@@ -0,0 +1,23 @@
+require 'rotp'
+require 'redis'
```
Comment (mdouchement) [inline spec/spec_helper.rb]: - Already required by `rotp-redis`

File: spec/spec_helper.rb [hunk 5]
```diff
@@ -0,0 +1,23 @@
...
+require 'redis-mutex'
```
Comment (mdouchement) [inline spec/spec_helper.rb]: - Already required by `rotp-redis`

File: spec/lib/rotp/opt_mem_caches/redis_spec.rb [hunk 6]
```diff
@@ -0,0 +1,67 @@
...
+          allow(subject).to receive(:retrieve).with(uid).and_return(nil)
+          allow(subject).to receive(:push).with(uid, otp)
+          allow_any_instance_of(RedisMutex).to receive(:lock).and_return(true)
+          allow_any_instance_of(RedisMutex).to receive(:unlock).and_return(true)
+
+          @opt_mem_cache_conf_backup = ROTP.config.opt_mem_cache_conf
+          ROTP.config.opt_mem_cache_conf[:expiration_time] = 300
+        end
+
+        subject { described_class.new }
```
Comment (mdouchement) [inline spec/lib/rotp/opt_mem_caches/redis_spec.rb]: - Implicitly done by RSpec

File: spec/lib/rotp/opt_mem_caches/redis_spec.rb [hunk 7]
```diff
@@ -0,0 +1,67 @@
...
+        end
+
+        it 'is used' do
+          expect(subject.unused?(uid, otp)).to eq(false)
+        end
+      end
+    end
+
+    describe '#push' do
+      let(:redis) { Object.new }
```
Comment (mdouchement) [inline spec/lib/rotp/opt_mem_caches/redis_spec.rb]: - How about `double`?

File: spec/spec_helper.rb [hunk 8]
```diff
@@ -0,0 +1,21 @@
+require 'rotp'
+require 'rotp-redis'
+
+ENV['RACK_ENV'] ||= 'test'
+
+ROOT_SPEC = File.join(File.dirname(__FILE__), '../')
+
+Dir[File.join(
```
Comment (mdouchement) [inline spec/spec_helper.rb]: RSpec should auto load all `*_spec.rb` files

File: lib/rotp-redis.rb [hunk 9]
```diff
@@ -0,0 +1,6 @@
+require 'rotp/opt_mem_caches/redis'
+
+module Rotp
```
Comment (mdouchement) [inline lib/rotp-redis.rb]: upcase

File: spec/spec_helper.rb (hunk 8 above)
Comment (HugoPouliquen) [inline spec/spec_helper.rb]: ?

It's the default spec file.

Comment (mdouchement) [inline spec/spec_helper.rb]: RSpec should work without this statement and the statement at line 6.

**Comment by HugoPouliquen** [comment]
@mdouchement PR is fixed

**Pull request merged**